File: Linear/Dynamic/Linked List/Singly Linked List/03_Deleting/Given_Position/Deleting.java
import java.util.Scanner;

class Node {
    int data;
    Node next;

    Node(int data) {
        this.data = data;
        this.next = null;
    }
}

class DeletingGivenPosition {
    public static Node deleteGivenNode(Node head, int pos) {
        Node temp = head;

        if (head == null || pos < 1)
        return null;

        // Head is to be deleted
        if (pos == 1) {
            head = head.next;
            temp = null;
            return head;
        }

        // Traverse to the node before the one to be deleted
        Node prev = null;
        for (int i = 1; i < pos && temp != null; i++) {
            prev = temp;
            temp = temp.next;
        }

        if (temp == null)
        return head;

        // Delete the node at the position
        prev.next = temp.next;

        return head;
    }

    public static void printList(Node head) {
        while (head != null) {
            System.out.print(head.data);
            if (head.next != null)
            System.out.print(" -> ");
            head = head.next;
        }

        System.out.println();
    }

    public static void main(String[] args) {
        Node head = new Node(1);
        head.next = new Node(2);
        head.next.next = new Node(3);
        head.next.next.next = new Node(4);

        System.out.println("Linked List before deletion");
        printList(head);

        Scanner sc = new Scanner(System.in);
        System.out.println("Enter a position: ");
        int pos = sc.nextInt();

        head = deleteGivenNode(head, pos);

        sc.close();
        
        System.out.println("Linked List after deletion");
        printList(head);
    }
}

File: Linear/Dynamic/Linked List/Doubly Linked List/03_Deleting/End/Deleting.js
class Node {
    constructor(data) {
        this.data = data;
        this.prev = null;
        this.next = null;
    }
}

// Function to remove the last node
function deleteLastNode(head) {
    /*Check if the list is empty or If the
    list has only one node, delete it*/
    if (head === null || head.next === null) {
        return null
    }

    // Find the last node
    let last = head;
    while (last.next !== null) {
        last = last.next;
    }

    // Delete the last node
    last.prev.next = null;
    last = null;

    return head;
}

function printList(head) {
    while (head !== null) {
        process.stdout.write(head.data.toString())
        if (head.next !== null)
            process.stdout.write(" <-> ")
        head = head.next
    }
    console.log()
}

// Create a hardcoded doubly linked list:
// 1 <-> 2 <-> 3
let head = new Node(1);
let second = new Node(2);
let third = new Node(3);

head.next = second;
second.prev = head;
second.next = third;
third.prev = second;

console.log("Linked List before deletion");
printList(head);

head = deleteLastNode(head);

console.log("Linked List after deletion");
printList(head);